File: ecos-records/src/main/java/ru/citeck/ecos/records2/source/dao/local/source/RecordsSourceRecordsDao.java
package ru.citeck.ecos.records2.source.dao.local.source;

import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.records3.record.dao.atts.RecordsAttsDao;
import ru.citeck.ecos.records3.record.atts.value.impl.EmptyAttValue;
import ru.citeck.ecos.records3.record.dao.query.RecordsQueryDao;
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery;
import ru.citeck.ecos.records3.record.dao.query.dto.res.RecsQueryRes;
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao;
import ru.citeck.ecos.records3.record.dao.RecordsDaoInfo;

import java.util.List;
import java.util.stream.Collectors;

public class RecordsSourceRecordsDao extends AbstractRecordsDao
                                     implements RecordsQueryDao,
                                                RecordsAttsDao {

    public static final String ID = "source";

    @Override
    public List<?> getRecordsAtts(List<String> records) {

        return records.stream().map(rec -> {
            RecordsDaoInfo info = recordsService.getSourceInfo(rec);
            if (info == null) {
                return EmptyAttValue.INSTANCE;
            }
            return info;
        }).collect(Collectors.toList());
    }

    @Override
    public RecsQueryRes<?> queryRecords(@NotNull RecordsQuery query) {
        return new RecsQueryRes<>(recordsService.getSourcesInfo());
    }

    @NotNull
    @Override
    public String getId() {
        return ID;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/dao/impl/source/RecordsSourceRecordsDao.kt
package ru.citeck.ecos.records3.record.dao.impl.source

import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.atts.value.impl.EmptyAttValue
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao
import ru.citeck.ecos.records3.record.dao.HasSourceIdAliases
import ru.citeck.ecos.records3.record.dao.atts.RecordsAttsDao
import ru.citeck.ecos.records3.record.dao.query.RecordsQueryDao
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import java.util.stream.Collectors

class RecordsSourceRecordsDao(services: RecordsServiceFactory) :
    AbstractRecordsDao(),
    RecordsQueryDao,
    RecordsAttsDao,
    HasSourceIdAliases {

    companion object {
        const val ID = "src"
        private val aliases = listOf("source")
    }

    private val recordsResolver = services.recordsResolver

    override fun getRecordsAtts(recordIds: List<String>): List<*> {

        return recordIds.stream().map { rec: String ->
            val info = recordsResolver.getSourceInfo(rec) ?: return@map EmptyAttValue.INSTANCE
            info
        }.collect(Collectors.toList())
    }

    override fun queryRecords(recsQuery: RecordsQuery): Any {
        return recordsResolver.getSourcesInfo()
    }

    override fun getId(): String {
        return ID
    }

    override fun getSourceIdAliases(): Collection<String> {
        return aliases
    }
}
